### PR TITLE
[inductor] Fix ldexp output dtype to match eager type promotion

### LINKED ISSUE #197091: [inductor] `torch.ldexp` output dtype differs from eager in both directions (int/bool × half → float32; half × float32 → half)
### 🐛 Describe the bug

**Summary.** `torch.ldexp` does not return the same dtype in eager and under Inductor: eager follows type promotion of both operands (bool or int mantissa with a half exponent → half; half mantissa with a float32 exponent → float32), while Inductor keeps the mantissa's dtype when it is floating point and uses float32 otherwise, so `bool × fp16` becomes float32 and `fp16 × fp32` becomes float16 (CPU and CUDA, torch 2.14.0 and nightly; `aot_eager` agrees with eager). A user sees no error; downstream dtype checks fail or precision is lost depending on which direction the rule went. Because the dtype is decided by the native lowering's propagation rule rather than by the operator's promotion semantics, the result differs from eager in both directions.

Inductor's native `ldexp` lowering decides the output dtype differently from eager (and from `aot_eager`, which agrees with eager):

```python
import torch

cases = {
    "bool  x fp16": (torch.ones(3, 4, dtype=torch.bool), torch.randn(3, 4).half()),
    "int64 x bf16": (torch.randint(1, 4, (3, 4)), torch.randn(3, 4).bfloat16()),
    "fp16  x fp32": (torch.randn(3, 4).half(), torch.randn(3, 4)),
    "fp16  x bf16": (torch.randn(3, 4).half(), torch.randn(3, 4).bfloat16()),
    "bf16  x fp64": (torch.randn(3, 4).bfloat16(), torch.randn(3, 4).double()),
}
for name, (a, b) in cases.items():
    torch._dynamo.reset()
    e = torch.ldexp(a, b).dtype
    i = torch.compile(torch.ldexp)(a, b).dtype
    print(f"{name}: eager {e} | inductor {i}")
```

Output (torch 2.14.0+cpu, same on 2.14.0+cu130 / Tesla T4 and on the 2026-09-11 nightly):

```
bool  x fp16: eager torch.float16  | inductor torch.float32
int64 x bf16: eager torch.bfloat16 | inductor torch.float32
fp16  x fp32: eager torch.float32  | inductor torch.float16
fp16  x bf16: eager torch.float32  | inductor torch.float16
bf16  x fp64: eager torch.float64  | inductor torch.bfloat16
```

Eager follows ordinary type promotion of the two operands (`ldexp(x, e) == x * 2**e`): an integer/bool mantissa with a half exponent gives half, and a half mantissa with a float32 exponent gives float32. Inductor instead keeps the mantissa's dtype when it is floating point and uses float32 when it is not — the exponent's dtype never takes part. `fp16 x fp16`, `fp32 x int64` etc. agree.

Both directions are user-visible: a model that keeps activations in bf16 and passes a float32 exponent gets a bf16 result after compile (precision loss), an integer mantissa with a half exponent gets float32 (memory / dtype mismatch further down).

Where: `torch/_inductor/lowering.py`, the `aten.ldexp` lowering added in #171721 and its `register_op_dtype_propagation_rules` entry; `aot_eager` (which uses the decomposition `input * pow(2, other)`) matches eager, so the decomposition's promotion is right and only the native lowering's rule is off.

Expected: `torch.compile(torch.ldexp)(a, b).dtype == torch.ldexp(a, b).dtype` for every dtype pair (i.e. `torch.result_type(a, b)` promoted to floating point).

### Error logs

No error; dtype mismatch only (values are numerically the same after casting).

### Versions

(torch 2.14.0+cpu, Windows 11, Python 3.14, MSVC 19.44 — `python -m torch.utils.collect_env` output below; the same reproducer was also run on the 2026-09-11 nightly wheel.)

```
PyTorch version: 2.14.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A
MSVC: Microsoft (R) C/C++ Optimizing Compiler Version 19.44.35228 for x64 (used by Inductor's C++ backend)

Python version: 3.14.7 (tags/v3.14.7:823f032, Aug  5 2026, 10:51:32) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Is XPU available: False
Is XNNPACK available: False

CPU:
Name: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
Manufacturer: GenuineIntel
Architecture: 9

Versions of relevant libraries:
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
[pip3] torchvision==0.29.0+cpu
[conda] Could not collect
```

Also reproduced on torch 2.14.0+cu130 (Kaggle, Tesla T4: `bool x fp16` → float32) and Linux CPU.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by SepehrRezaee**
> **Disclosure (PyTorch AI policy):** This PR and its description were drafted with an AI coding assistant; the diff and this text have been reviewed by the human submitter, and the PR is opened in **draft** for maintainer review.

## Description

Under `torch.compile` (Inductor), `torch.ldexp` returns a different dtype than eager for mixed-dtype inputs, in both directions (e.g. `fp16 x fp32`: eager `float32`, Inductor `float16`; `int64 x bf16`: eager `bfloat16`, Inductor `float32`). A user sees no error - downstream dtype checks fail or precision silently changes (#197091).

Root cause: the Inductor `ldexp` lowering has two paths. The native path (float mantissa, integral exponent) is fine. The fallback path (`x * pow(2, n)`, taken whenever the mantissa is not floating point or the exponent is not integral) hardcoded `out_dtype = float32 if is_integer_type(x) else x_dtype` - the exponent's dtype never participates.

Eager computes `mul(x, _pow2(x, n))` (`aten/src/ATen/native/BinaryOps.cpp`), where `_pow2`'s dtype is derived from both operands, so the effective output dtype is the type promotion of both operands, promoted to a floating dtype when that promotion is integral.

Fix: compute the fallback output dtype as `torch.promote_types(x_dtype, n_dtype)`, falling back to the default floating dtype when the promotion is integral. The native path and its dtype propagation rule are unchanged.

## Test Plan

Added `test_ldexp_dtype_promotion` to `test/inductor/test_op_dtype_prop.py`: it checks compiled-vs-eager dtype equality (and value agreement) on the mixed-dtype pairs from the issue - `(bool, fp16)`, `(int64, bf16)`, `(fp16, fp32)`, `(fp16, bf16)`, `(bf16, fp64)`.

```
python test/inductor/test_op_dtype_prop.py -v -k ldexp
```

Not run locally (no build environment available); left to CI.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by SepehrRezaee** [comment]
@codex
